### PR TITLE
[Core] Residual criteria and block B&S correction

### PR DESCRIPTION
**Pull request by loumalouomega**
In this PR I do what @RiccardoRossi proposes in #4272. Unfortunately doesn't solve the issue completely. We will need to do a convergence check for the TL (and UL) element. The SD converges in 2 iterations.

### CODE REVIEW

**Comment by loumalouomega** [comment]
Looks like it fails due to a small tolerance

**Comment by RiccardoRossi** [comment]
i committed some small modififications to this branch, in order to conditionally compute the RHS just in case it is needed by the convergence criteria.

i also try to fix the problem with MPCs by skipping master and slave dofs in the norm computation.

Still the test of @rubenzorrilla does not converge properly even if i use the SmallDisplacementElement and a static analysis. I am convinced that the problem comes out of an improper linearization of the follower pressure. Still investigating, nevertheless @marandra could u try out this branch?

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h [hunk 1]
```diff
@@ -210,19 +206,48 @@ class ResidualCriteria
...
+        }
+        
+        for(auto& mpc : rModelPart.MasterSlaveConstraints())
+        {
+            auto& slave_dofs = mpc.GetSlaveDofsVector();
+            auto& master_dofs = mpc.GetMasterDofsVector();
+
+            for(auto& dof : slave_dofs)
+                mActiveDofs[dof->EquationId()] = false;
+            for(auto& dof : master_dofs)
```
Comment (philbucher) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h]: Why do you exclude the master dofs?

Or why excluding the mpcs at all?

Comment (RiccardoRossi) [reply kratos/solving_strategies/convergencecriterias/residual_criteria.h]: the residual on the dirichlet dofs and on the mpcs is non zero (it converges to the reactions), what goes to zero is only the norm of the residual on the non dirichlet (and nonmaster/slave nodes)

**Comment by RiccardoRossi** [comment]
just to add that i think this correction is quite important, and that we should eventually backport it even to the release

**Comment by loumalouomega** [comment]
Test is not working due to tolerance issues

**Comment by RiccardoRossi** [comment]
about the test i think you are right, nevertheless the fix is relevant

**Comment by philbucher** [comment]
the test is failing due to some other reason, it is temporarily commented (already in master)
will merge master here now then we see what happens

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h (hunk 1 above)
Comment (philbucher) [reply kratos/solving_strategies/convergencecriterias/residual_criteria.h]: ok

FYI @adityaghantasala

**Comment by loumalouomega** [comment]
I tried with last updates. The truss and beam tests are still failing @KlausBSautter and @philbucher

**Comment by philbucher** [comment]
which ones exactly?

**Comment by loumalouomega** [comment]
Most of the truss patch test (among other ones)

**Comment by philbucher** [comment]
is it the tests that use NR? It could be a matter of choosing a different tolerance then
How much do the results differ?

**Comment by loumalouomega** [comment]
> is it the tests that use NR? It could be a matter of choosing a different tolerance then
> How much do the results differ?

Yep. A lot, some of them give 0 instead of different results

**Comment by philbucher** [comment]
hm that sounds strange, are you sure the changes in this PR are all correct?

Travis is failing bcs of a more serious error, maybe you check this first?

**Comment by loumalouomega** [comment]
I will check, but doesn't happen in my computer

**Comment by loumalouomega** [comment]
> I will check, but doesn't happen in my computer

It is just an error in the results

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h [hunk 2]
```diff
@@ -385,6 +404,8 @@ class ResidualCriteria
 
     TDataType mReferenceDispNorm;   /// The norm at the beginning of the iterations
 
+    std::vector<bool> mActiveDofs;
```
Comment (loumalouomega) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h]: I don't remember why I added this

**Comment by philbucher** [comment]
this happens in Travis:
~~~
/home/travis/build/KratosMultiphysics/Kratos/applications//StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
............................................Check failed in file /usr/include/boost/numeric/ublas/storage.hpp at line 190:
i < size_
terminate called after throwing an instance of 'boost::numeric::ublas::bad_index'
  what():  bad index
~~~

Thats what I meant you should first take a look

**Comment by RiccardoRossi** [comment]
i added it.

you need to put a 1 on the diaginal for inactive ones

**Comment by loumalouomega** [comment]
Now it only does that computation when MPC (avoiding extra computation cost for regular cases)

**Comment by philbucher** [comment]
Travis still fails

**Comment by loumalouomega** [comment]
> Travis still fails

Last commit was not to prevent Travis fails, but not to penalize the residual computation when there are not MPC present in the model part.

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h [hunk 3]
```diff
@@ -325,18 +353,35 @@ class ResidualCriteria
...
+            #pragma omp parallel for firstprivate(residual_dof_value) reduction(+:residual_solution_norm, dof_num)
+            for (int i = 0; i < number_of_dof; i++) {
+                auto it_dof = it_dof_begin + i;
+
+                if (!it_dof->IsFixed()) {
+                    const IndexType dof_id = it_dof->EquationId();
+                    residual_dof_value = TSparseSpace::GetValue(b,dof_id);
+                    residual_solution_norm += std::pow(residual_dof_value, 2);
+                }
                 dof_num++;
```
Comment (philbucher) [inline kratos/solving_strategies/convergencecriterias/residual_criteria.h]: shouldn't this also be in the loop?

**Review by philbucher — COMMENTED** [review]
do these changes also have to be ported to the displacement-criterion?

File: kratos/solving_strategies/convergencecriterias/residual_criteria.h (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/convergencecriterias/residual_criteria.h]: you are probably right, although i do not completely understand the use of it

Comment (philbucher) [reply kratos/solving_strategies/convergencecriterias/residual_criteria.h]: used here:
https://github.com/KratosMultiphysics/Kratos/blob/1ca3a824725fa4a727ce649bbb3bf7eb76678760/kratos/solving_strategies/convergencecriterias/residual_criteria.h#L177-L178

If it is outside the loop then it is incorrect

Comment (philbucher) [reply kratos/solving_strategies/convergencecriterias/residual_criteria.h]: fixed it

**Comment by loumalouomega** [comment]
I adjusted tolerances of the tests, now all the tests pass on FullDebug

**Review by philbucher — CHANGES_REQUESTED** [review]
you cannot just adjust aka decrease the tolerances!
If there is still an issue then we should update the values no lower the tolerance

Otherwise if in the future sth small changes it might not be detected with the lower tolerances?

Is everything else working now? Then we can have a look and update the values

**Comment by loumalouomega** [comment]
> you cannot just adjust aka decrease the tolerances!
> If there is still an issue then we should update the values no lower the tolerance
> 
> Otherwise if in the future sth small changes it might not be detected with the lower tolerances?
> 
> Is everything else working now? Then we can have a look and update the values

The differences are small (the difference is on the last 4 decimals in numbers of the 10^4 order), in any case you are right on that, in that case we should wait for the responsible of the tests to give green light

**Comment by loumalouomega** [comment]
Ping @KlausBSautter

**Comment by loumalouomega** [comment]
> > you cannot just adjust aka decrease the tolerances!
> > If there is still an issue then we should update the values no lower the tolerance
> > Otherwise if in the future sth small changes it might not be detected with the lower tolerances?
> > Is everything else working now? Then we can have a look and update the values
> 
> The differences are small (the difference is on the last 4 decimals in numbers of the 10^4 order), in any case you are right on that, in that case we should wait for the responsible of the tests to give green light

We should add a relative error to the unittest suite ((value-reference)/reference) < tolerance.

**Comment by RiccardoRossi** [comment]
as a comment, the overall effect of this PR should be to reduce the number of iterations to DETECT convergence when using the residual criterion.

essentially the problem was that before this commit instead of detecting as converged when

r(i+1)/r(0) < tol

it was detecting convergence when

r(i+1)/r(1) < tol

in the practice this was implying requiring a much tighter convergence criterion than it is done now.

**Comment by philbucher** [comment]
just talked to @KlausBSautter , he fixed the issues now

**Comment by philbucher** [comment]
@RiccardoRossi doe we need to apply the same corrections also on the displacement-criterion?

**Comment by RiccardoRossi** [comment]
no, since the displacement criteria does

( d(i+1)-d(i) ) / d(i+1) < tol

that criteria does not need the recomputation of the residual so i would say the current version is correct

**Comment by philbucher** [comment]
ok thx

**Comment by loumalouomega** [comment]
Do we revert then the tolerance change?

**Comment by philbucher** [comment]
> Do we revert then the tolerance change?

Yes, will fix the tests tmr

**Comment by RiccardoRossi** [comment]
@philbucher @KlausBSautter just to be sure: you are going to fix the failing test in this PR, correct?

if so, i leave to you approving this. Once done i would also vote for backporting this bugfix to the release

**Comment by philbucher** [comment]
Yes, in this PR, by later today

**Comment by KlausBSautter** [comment]
like that? @philbucher we can talk about this later

**Review by philbucher — APPROVED** [review]
Thanks!

This does not break the trilinos-app right?

As general info, I finalizing the generic convergence criteria atm, stay tuned ;)

**Comment by loumalouomega** [comment]
> Thanks!
> 
> This does not break the trilinos-app right?
> 
> As general info, I finalizing the generic convergence criteria atm, stay tuned ;)

I already fixed the Trilinos App

**Pull request merged**